### PR TITLE
Start of the Exchange

### PR DESCRIPTION
**Pull request by Wobin**
Also including an auto setup of the localtunnel, if you set the env variable for the path to the DTAuth mod.
Moved some of the schema from the darktide.server to the schema.server

### CODE REVIEW

File: app/data/schemas.server.ts [hunk 1]
```diff
@@ -142,3 +142,436 @@ export const BlessingSchema = z
...
+
+export const JoinSchema = z.object({
+  queuePosition: z.number(),
+  queueTicket: z.string(),
+  retrySuggestion: z.number(),
+})
+
+
+export const CharactersSchema = z.object({
+  _links: z.object({ self: z.object({ href: z.string() }) }),
```
Comment (danreeves) [inline app/data/schemas.server.ts:275]: FYI I've been ignoring _links in my schema

File: app/data/schemas.server.ts [hunk 2]
```diff
@@ -142,3 +142,436 @@ export const BlessingSchema = z
...
+  characters: z.array(
+    z.union([
```
Comment (danreeves) [inline app/data/schemas.server.ts:277]: Kinda hard to tell with such a big object, what the difference between the character types in this union?

File: app/root.tsx [hunk 3]
```diff
@@ -14,8 +14,11 @@ import tailwind from "~/tailwind.css"
 import { authenticator } from "~/services/auth.server"
 
 // Sets up Cron singletons to perform timed jobs on the server
-import { initJobs } from "~/jobs/index.server"
+import { initJobs, getTunnel } from "~/jobs/index.server"
 initJobs && initJobs()
+if( process.env.NODE_ENV !== "production") {
```
Comment (danreeves) [inline app/root.tsx:19]: It's be nice if this check was internal to the getTunnel function

File: app/jobs/index.server.ts [hunk 4]
```diff
@@ -15,3 +16,14 @@ export function initJobs() {
       return true
     })
 }
+
+export async function getTunnel(){
+  let soon = new Date()
```
Comment (danreeves) [inline app/jobs/index.server.ts:21]: It'd be nice to wrap this in a check that the DTAUTHDATA_PATH env var exists

File: app/routes/armoury/$character/exchange.tsx [hunk 5]
```diff
@@ -1,5 +1,101 @@
...
+            </div>
+			{item.item.description.overrides.perks.length > 0 ?
+			<div>
+				<div>Perks</div>
+				{ item.item.description.overrides.perks.map(perk => (
+					<div>{perk.id}</div>
+				))}
+				<span></span>
+			</div>
+			:""}
```
Comment (danreeves) [inline app/routes/armoury/$character/exchange.tsx:89]: you can use null when you don't have something to render

File: app/services/darktide.server.ts [hunk 6]
```diff
@@ -301,3 +188,23 @@ export async function getAccountTrait(auth: AuthToken, traitCategory: string) {
     }
   }
 }
+export async function getShopFor(auth: AuthToken, archetype: string, id: string) {
```
Comment (danreeves) [inline app/services/darktide.server.ts:191]:
```suggestion
export async function getCharacterStore(auth: AuthToken, characterArchetype: string, characterId: string) {
```
rename for consistency and clarity

fyi this function will get reused for melks store where the url changes from `credits_store` to something else `marks_store`. At that point the function arguments should become an object like 

```ts
type GetCharacterStoreArgs = {
	auth: AuthToken
	characterArchetype: string
	characterId: string
	store: "credits" | "marks"
}
export async function getCharacterStore({auth, characterArchetype, characterId, store}: GetCharacterStoreArgs)
```

**Review by danreeves — COMMENTED** [review]
Some comments and requests. Lemme know if you want to work on them now or merge and work on them progressively

File: app/data/authtoken.server.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import type { AuthToken } from "@prisma/client"
-import { prisma } from "~/data/db.server"
+import { prisma } from "../data/db.server"
```
[the remark this answers is not in the document]
Comment (Wobin) [reply app/data/authtoken.server.ts:2]: yeah, it really doesn't like the ~/ path for some reason, I figure if we write more tests we can convert as we go, unless there's a base setting I don't know of

[the remark this answers is not in the document]
Comment (danreeves) [reply app/data/authtoken.server.ts:2]: Yeah, vitest probably has a configuration for it

[the remark this answers is not in the document]
Comment (danreeves) [reply app/data/authtoken.server.ts:2]: https://stackoverflow.com/questions/72468249/vitest-src-folder-alias-not-resolved-in-test-files

File: app/data/schemas.server.ts (hunk 2 above)
Comment (Wobin) [reply app/data/schemas.server.ts]: I was auto converting the json without actually checking the format

**Review by danreeves — APPROVED** [review]

**Pull request merged**